### PR TITLE
Broccoli-based builds

### PR DESCRIPTION
**Pull request by davewasmer**
This PR significantly changes and simplifies the CLI commands that require a build of the app (i.e. `build`, `serve`, `routes`, `console`, `test`).

First, it introduces a Project class to represent the app at build time. This consolidates the build related code that was duplicated across several command classes and provides a single consistent interface.

Second, it converts the ad-hoc build processes to a Broccoli based build pipeline. Apps and addons can add a `denali-build.js` to their root directory, which should export a function that accepts a Broccoli tree and returns it. This allows apps and addons to customize their own build process. I.e. you want to write you app in Typescript? Just return the supplied tree wrapped in a Typescript compiler tree.

### CODE REVIEW

**Comment by davewasmer** [comment]
@knownasilya @acorncom thoughts?

**Comment by knownasilya** [comment]
B-E-A-U-T-I-F-U-L. I especially like the use of `dedent`. On another note it seems some broccoli deps are missing like merge trees and funnel.

File: lib/cli/lib/project.js [hunk 1]
```diff
@@ -0,0 +1,227 @@
+import fs from 'fs';
+import path from 'path';
+import nsp from 'nsp';
+import dedent from 'dedent-js';
+import escape from 'js-string-escape';
+import broccoli from 'broccoli';
+import { Watcher } from 'broccoli/lib';
+import rimraf from 'rimraf';
+import MergeTree from 'broccoli-merge-trees';
```
Comment (knownasilya) [inline lib/cli/lib/project.js]: missing in package?

File: lib/cli/lib/project.js [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+import Funnel from 'broccoli-funnel';
```
Comment (knownasilya) [inline lib/cli/lib/project.js]: missing in package?

File: lib/cli/lib/project.js [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import fs from 'fs';
+import path from 'path';
+import nsp from 'nsp';
+import dedent from 'dedent-js';
+import escape from 'js-string-escape';
+import broccoli from 'broccoli';
+import { Watcher } from 'broccoli/lib';
```
Comment (knownasilya) [inline lib/cli/lib/project.js]: `broccoli` missing in package?

File: lib/cli/lib/project.js [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+import rimraf from 'rimraf';
+import MergeTree from 'broccoli-merge-trees';
+import Funnel from 'broccoli-funnel';
+import BabelTree from 'broccoli-babel-transpiler';
+import LintTree from 'broccoli-lint-eslint';
+import { sync as copyDereferenceSync } from 'copy-dereference';
```
Comment (knownasilya) [inline lib/cli/lib/project.js]: missing?

File: lib/cli/lib/project.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+import tmp from 'tmp';
```
Comment (knownasilya) [inline lib/cli/lib/project.js]: missing?

**Comment by acorncom** [comment]
That looks really nice :+1:

**Comment by davewasmer** [comment]
@knownasilya good catch - had those installed locally but not declared in package.json. Add those in now. This will likely impact the addon blueprints PR, but I purposefully left that alone for now (one big change at a time :wink:).

**Comment by knownasilya** [comment]
I did rebase the addons branch against current master so it should be easier to rebase against this, hopefully.

**Pull request merged**